src/TodosContext.tsx

```diff
@@ -0,0 +1,271 @@
+  const [isTitleFieldFocused, setIsTitleFieldFocused] = useState(true);
+  const [error, setError] = useState('');
+
+  const filteredTodos: Todo[] = useMemo(() => todos.filter(({ completed }) => {
+    switch (filter) {
+      case Filter.Active:
+        return !completed;
+      case Filter.Completed:
+        return completed;
+      default:
+        return true;
+    }
+  }), [filter, todos, tempTodos]);
+
+  const timerIdRef = useRef<number>(0);
+
+  useEffect(() => {
+    if (timerIdRef.current) {
+      window.clearTimeout(timerIdRef.current);
+    }
+
+    timerIdRef.current = window.setTimeout(() => {
+      setError(ErrorsMessages.None);
+    }, 3000);
+
+    return () => clearTimeout(timerIdRef.current);
+  }, [error]);
+
+  const loadTodos = async () => {
+    try {
+      setError(ErrorsMessages.None);
+      const fetchedTodos = await getTodos(USER_ID);
+
+      setTodos(fetchedTodos);
+    } catch {
+      setError(ErrorsMessages.Load);
+    }
+  };
+
+  useEffect(() => {
+    loadTodos();
+  }, []);
+
+  const handlerTitleFieldFocused = (status: boolean) => {
+    setIsTitleFieldFocused(status);
+  };
+
+  const handleFilter = (value: Filter) => {
+    setFilter(value);
+  };
+
+  const handleClearError = () => {
+    setError(ErrorsMessages.None);
+  };
+
+  const isAllTodosCompleted = todos.every(todo => todo.completed);
+
+  const handleToggleAllTodos = async () => {
+    const updatedTodos = [...todos];
+    const newStatus = { completed: false };
+
+    if (!isAllTodosCompleted) {
+      newStatus.completed = true;
+    }
+
+    const todosToRequest = isAllTodosCompleted
+      ? updatedTodos
+      : updatedTodos.filter((todo) => !todo.completed);
+
+    try {
+      setError(ErrorsMessages.None);
+      setTempTodos(todosToRequest);
+
+      const updatedTodosResults = await Promise.all(
+        todosToRequest.map(async todo => {
+          return await updateTodo(todo.id, newStatus) as Todo;
+        }),
+      );
+
+      updatedTodosResults.forEach((updatedTodo) => {
+        const indexOfUpdatedTodo = updatedTodos
+          .findIndex(({ id }) => id === updatedTodo.id);
+
+        updatedTodos[indexOfUpdatedTodo].completed = updatedTodo.completed;
+      });
+
+      setTodos(updatedTodos);
+    } catch {
+      setError(ErrorsMessages.Update);
+    } finally {
+      setTempTodos([]);
+    }
+  };
+
+  const handleAddTodo = async (title: string): Promise<boolean | undefined> => {
+    if (!title) {
+      setError(ErrorsMessages.Empty);
+
+      return false;
+    }
+
+    const newTodo = {
+      id: 0,
+      userId: USER_ID,
+      title,
+      completed: false,
+    };
+
+    try {
+      setError(ErrorsMessages.None);
+      setTempTodos([newTodo]);
+      const addedTodo = await addTodo(newTodo);
+
+      setTodos(prevTodos => [...prevTodos, addedTodo]);
+
+      return true;
+    } catch {
+      setError(ErrorsMessages.Add);
+
+      return false;
```

Comment:
```suggestion
      setTodos(prevTodos => [...prevTodos, addedTodo]);
    } catch {
      setError(ErrorsMessages.Add);
```
It's strange to return true or false in handler

Reply:
this is necessary in order to know whether our promise was fulfilled normally or with an error. I don't know how to do it differently, but it works)